packages/api/src/webhooks/cannon.ts

```diff
@@ -0,0 +1,253 @@
+// import { getWebhooks } from "../controllers/helpers";
+import Model from "../store/model";
+import { fetchWithTimeout, fetchWithTimeoutAndSleep } from "../util";
+import logger from "../logger";
+import uuid from "uuid/v4";
+
+import { parse as parseUrl } from "url";
+import bytesToUuid from "uuid/lib/bytesToUuid";
+const isLocalIP = require("is-local-ip");
+const { Resolver } = require("dns").promises;
+// const resolver = new Resolver();
+
+const WEBHOOK_TIMEOUT = 5 * 1000;
+const MAX_BACKOFF = 10 * 60 * 1000;
+const BACKOFF_COEF = 1.2;
+
+export default class WebhookCannon {
+  db: DB;
+  store: Model;
+  running: boolean;
+  verifyUrls: boolean;
+  resolver: any;
+  constructor({ db, store, verifyUrls }) {
+    this.db = db;
+    this.store = store;
+    this.running = true;
+    this.verifyUrls = verifyUrls;
+    this.resolver = new Resolver();
+    // this.start();
+  }
+
+  async start() {
+    console.log("WEBHOOK CANNON STARTED");
+    if (this.running) {
+      await this.db.queue.setMsgHandler(this.processEvent.bind(this));
+    }
+  }
+
+  async processEvent(msg: Notification) {
+    console.log("EVENT TRIGGERED ON THE WEBHOOK");
+    let event = await this.db.queue.pop(this.onTrigger.bind(this));
+    console.log("event: ", event);
+  }
+
+  stop() {
+    this.db.queue.unsetMsgHandler();
+    this.running = false;
+  }
+
+  disableUrlVerify() {
+    this.verifyUrls = false;
+  }
+
+  calcBackoff(lastInterval): number {
+    if (!lastInterval || lastInterval < 1000) {
+      lastInterval = 5000;
+    }
+    let newInterval = lastInterval * BACKOFF_COEF;
+    if (newInterval > MAX_BACKOFF) {
+      return lastInterval;
+    }
+    return newInterval;
+  }
+
+  retry(event) {
+    event.lastInterval = this.calcBackoff(event.lastInterval);
+    event.status = "pending";
+    event.retries = event.retries ? event.retries + 1 : 1;
+    this.db.queue.updateMsg(event);
+  }
+
+  async _fireHook(
+    event: Queue,
+    webhook: Webhook,
+    sanitized: Stream,
+    user: User,
+    verifyUrl = true
+  ) {
+    console.log(`trying webhook ${webhook.name}: ${webhook.url}`);
+    let ips, urlObj, isLocal;
+    if (verifyUrl) {
+      try {
+        urlObj = parseUrl(webhook.url);
+        if (urlObj.host) {
+          ips = await this.resolver.resolve4(urlObj.hostname);
+        }
+      } catch (e) {
+        console.error("error: ", e);
+        throw e;
+      }
+    }
+
+    // This is mainly useful for local testing
+    if (user.admin || verifyUrl === false) {
+      isLocal = false;
+    } else {
+      try {
+        if (ips && ips.length) {
+          isLocal = isLocalIP(ips[0]);
+        } else {
+          isLocal = true;
+        }
+      } catch (e) {
+        console.error("isLocal Error", isLocal, e);
+        throw e;
+      }
+    }
+    if (isLocal) {
+      // don't fire this webhook.
+      console.log(
+        `webhook ${webhook.id} resolved to a localIP, url: ${webhook.url}, resolved IP: ${ips}`
+      );
```

Comment:
**nit:** Maybe `return` here so we reduce nesting without the need for the `else`?